src/platform/unix/mod.rs

```diff
@@ -465,19 +464,23 @@ impl OsIpcReceiverSet {
                                 shared_memory_regions));
                     }
                     Err(err) if err.channel_is_closed() => {
-                        hangups.insert(pollfd.fd);
-                        selection_results.push(OsIpcSelectionResult::ChannelClosed(
-                                    *self.fdids.get(&pollfd.fd).unwrap()))
+                        let id = self.fdids.remove(&pollfd.fd).unwrap();
+                        unsafe {
+                            libc::close(pollfd.fd);
+                        }
+                        selection_results.push(OsIpcSelectionResult::ChannelClosed(id))
                     }
                     Err(err) => return Err(err),
                 }
                 pollfd.revents = pollfd.revents & !POLLIN
             }
         }
 
-        if !hangups.is_empty() {
-            self.pollfds.retain(|pollfd| !hangups.contains(&pollfd.fd));
-        }
+        // Avoid the use of `self` in closue to avoid errors due to the
+        // mutable borrow of `self` at line 483
+        let ids = &self.fdids;
```

Comment:
BTW, just for explanation (in case it isn't clear): the issue here isn't the borrow checker being dumb, but rather it's about the way how closures are handled. They automatically close over any binding used within the closure's code -- in this case the binding is `self`. The compiler doesn't try to figure out that you are actually accessing only one sub-field. If you want only the sub-field to be closed over, you have to create an explicit binding for it... Which is precisely what you did :-)